src/client/datascience/notebook/notebookControllerManager.ts

```diff
@@ -259,58 +256,49 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
             }
         });
 
-        // Map KernelConnectionMetadata => NotebookController
-        const allControllers: VSCodeNotebookController[] = [];
         connectionsWithLabel.forEach((value) => {
-            const controllers = this.createNotebookController(value.connection, value.label);
-            if (controllers) {
-                allControllers.push(...controllers);
-            }
+            this.createNotebookController(value.connection, value.label);
         });
-
-        return allControllers;
     }
-    private createNotebookController(
-        kernelConnection: KernelConnectionMetadata,
-        label: string
-    ): VSCodeNotebookController[] | undefined {
+    private createNotebookController(kernelConnection: KernelConnectionMetadata, label: string) {
         try {
             // Create notebook selector
-            return [
+            [
                 [kernelConnection.id, JupyterNotebookView],
                 [`${kernelConnection.id} (Interactive)`, InteractiveWindowView]
-            ].map(([id, viewType]) => {
-                const controller = new VSCodeNotebookController(
-                    kernelConnection,
-                    id,
-                    viewType,
-                    label,
-                    this.notebook,
-                    this.commandManager,
-                    this.kernelProvider,
-                    this.preferredRemoteKernelIdProvider,
-                    this.context,
-                    this.pathUtils,
-                    this.disposables,
-                    this.languageService,
-                    this.workspace,
-                    this.isLocalLaunch ? 'local' : 'remote',
-                    this.interpreterPackages,
-                    this.configuration,
-                    this.widgetCoordinator
-                );
-                // Hook up to if this NotebookController is selected or de-selected
-                controller.onNotebookControllerSelected(
-                    this.handleOnNotebookControllerSelected,
-                    this,
-                    this.disposables
-                );
-
-                // We are disposing as documents are closed, but do this as well
-                this.disposables.push(controller);
-
-                return controller;
-            });
+            ]
+                .filter(([id]) => !this.registeredControllers.has(id))
```

Comment:
This way we can try to create the controllers without having to worry about duplicates.